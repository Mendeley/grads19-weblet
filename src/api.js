import axios from "axios";
import { toast } from "react-toastify";

export const getConferenceList = async () => {
  const response = await axios.get("http://localhost:8080/conferences");
  return response.data;
};

export const getConferenceById = async id => {
  const response = await axios.get(`http://localhost:8080/conferences/${id}`);
  return response.data;
};

export const createNewConference = async conference => {
  await axios.post("http://localhost:8080/conferences", conference);
};

export const deleteConferenceById = async id => {
  try {
    await axios.delete(`http://localhost:8080/conferences/${id}`);
    toast.success("Conference successfully deleted!");
  } catch (error) {
    toast.error("Conference not deleted!");
    console.log(error);
    throw error;
  }
};

export const updateConferenceById = async (id, conference) => {
  try {
    await axios.patch(`http://localhost:8080/conferences/${id}`, conference);
    toast.success("Conference details have been updated!");
  } catch (error) {
    toast.error("Conference details have not been updated!");
    console.log(error);
    throw error;
  }
};

export const createNewUser = async user => {
  try {
    await axios.post("http://localhost:8080/users", user);
    toast.success("User successfully registered!");
  } catch (error) {
    toast.error("User registration failed!");
    console.log(error);
    throw error;
  }
};

export const loginUser = async user => {
  try {
    const response = await axios.post("http://localhost:8080/auth/login", user);
    toast.success("Login successful!");
    return response.data;
  } catch (error) {
    toast.error("Login failed!");
    console.log(error);
    throw error;
  }
};

export const getUserById = async (id, token) => {
  const response = await axios.get(`http://localhost:8080/users/${id}`, {
    headers: { Authorization: token }
  });
  console.log(response)
  return response.data;
};

export const logoutUser = async token => {
  try {
    await axios.delete("http://localhost:8080/auth/logout", {
      headers: { Authorization: token }
    });
    toast.success("Logout successful!");
  } catch (error) {
    toast.error("Logout failed!");
    console.log(error);
    throw error;
  }
};

export const updateUserById = async (id, user, token) => {
  try {
    await axios.patch(
      `http://localhost:8080/users/${id}`, 
      user,
      { headers: { Authorization: token}}
      );
  toast.success("Profile details have been updated!");
  } catch (error) {
  toast.error("Profile details have not been updated!");
  throw error;
  }
  };