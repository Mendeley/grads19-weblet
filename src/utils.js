export const getTimestring = date =>
  `${date.getHours()}:${
    date.getMinutes() < 10 ? "0" + date.getMinutes() : date.getMinutes()
  }`;

export const getDatestring = date =>
  `${date.getDate()}/${date.getMonth() + 1}/${date.getFullYear()}`;
