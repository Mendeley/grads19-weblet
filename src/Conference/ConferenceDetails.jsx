import React, { useEffect, useState } from "react";
import styled from "styled-components";
import { getTimestring, getDatestring } from "../utils";
import { Link, useParams } from "react-router-dom";
import { getConferenceById } from "../api.js";

const StyledConferenceDetails = styled.div`
  height: 400px;
  padding: 20px;
`;

const StyledDetailsCardHeading = styled.h3`
  background: #dbd8db;
  width: 100%;
  height: 55px;
  font-size: 30px;
  font-weight: bold;
  text-align: center;
  vertical-align: middle;
`;

const StyledDetailsCard = styled.div`
  border-style: solid;
  border-color: black;
  border-radius: 25px;
  background: #dbd8db;
  width: 65%;
  height: 370px;
  transition: 0.3s;
  :hover {
    box-shadow: 0 8px 16px 0 rgba(0, 0, 0, 0.2);
  }
  vertical-align: middle;
  margin: 0 auto;
`;

export const StyledLink = styled(Link)`
  color: #7a517d;
`;

const ConferenceDetails = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [conference, setConference] = useState(null);
  const [errorCaught, setErrorCaught] = useState(false);
  const { id } = useParams();

  const fetchData = async conferenceId => {
    setIsLoading(true);

    try {
      const conference = await getConferenceById(conferenceId);
      setConference(conference);
    } catch (error) {
      setErrorCaught(true);
    }
    setIsLoading(false);
  };

  useEffect(() => {
    if (id) {
      fetchData(id);
    }
  }, [id]);

  if (isLoading) {
    return <p>Loading...</p>;
  }

  if (errorCaught) {
    return <p>An error has occurred...</p>;
  }

  const { name, topic, dateTime, city, description } = conference || {};
  const date = new Date(dateTime);

  return (
    <StyledConferenceDetails>
      <StyledDetailsCard>
        <StyledDetailsCardHeading>{name}</StyledDetailsCardHeading>
        <p>{topic}</p>
        <p>{getDatestring(date)}</p>
        <p>{getTimestring(date)}</p>
        <p>{city}</p>
        <p>{description}</p>
        <StyledLink to={`/${id}/edit`}>Edit Conference</StyledLink>
      </StyledDetailsCard>
    </StyledConferenceDetails>
  );
};
export default ConferenceDetails;
