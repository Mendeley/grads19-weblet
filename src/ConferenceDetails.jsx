import React from "react";
// import Button from "./Button.jsx";
import styled from "styled-components";
import { getTimestring, getDatestring } from "./utils";

const StyledConferenceDetails = styled.div`
  width: 100%;
  height: 400px;
  color: white;
  background: darkslategray;
`;

const ConferenceDetails = ({ conference, setConferenceDetailsOpen }) => {
  const { name, topic, dateTime, city, description } = conference;
  const date = new Date(dateTime);

  return (
    <StyledConferenceDetails>
      <h3>{name}</h3>
      <p>{topic}</p>
      <p>{getDatestring(date)}</p>
      <p>{getTimestring(date)}</p>
      <p>{city}</p>
      <p>{description}</p>
      <button onClick={() => setConferenceDetailsOpen(false)}>Back</button>
    </StyledConferenceDetails>
  );
};
export default ConferenceDetails;
