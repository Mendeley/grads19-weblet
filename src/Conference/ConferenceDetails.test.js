import ConferenceDetails from "./ConferenceDetails";
import { configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Router } from "react-router-dom";
import React from "react";
import { getConferenceById } from "../api";
import { createMemoryHistory } from "history";
import { act } from "react-dom/test-utils";
import { getTimestring, getDatestring } from "../utils";

jest.mock("react-router-dom", () => {
  const originalReactRouter = jest.requireActual("react-router-dom");
  return {
    ...originalReactRouter,
    useParams: jest.fn(() => ({ id: "1" })),
    //useHistory: jest.fn(),
    Link: () => "hey I'm a link"
  };
});

jest.mock("../api", () => ({
  getConferenceById: jest.fn()
}));

configure({ adapter: new Adapter() });

/* This is a temporary workaround for enzyme/react bug #14769 (github.com/facebook/react/issues/14769):

Enzyme throws an warning when used with useEffect hook. We're surpressing this warning as it's not informative
and will be removed in future releases.
*/
const consoleError = console.error;
beforeAll(() => {
  jest.spyOn(console, "error").mockImplementation((...args) => {
    if (
      !args[0].includes(
        "Warning: An update to %s inside a test was not wrapped in act"
      )
    ) {
      consoleError(...args);
    }
  });
});

afterEach(() => {
  jest.clearAllMocks();
});

describe("ConferenceDetails", () => {
  const mockData = {
    id: 1,
    name: "Festival of Marketing",
    dateTime: "2019-11-12T12:34:11Z",
    city: "London",
    description:
      "From Festivalofmarketing.com: The Festival of Marketing is a unique experience where ambitious marketers can discover, learn, celebrate and shape the future together. As the largest global event dedicated to brand marketers, the Festival reflects the very nature of ...",
    topic: "Marketing"
  };

  it("renders conference details", async () => {
    expect.assertions(6);

    const history = createMemoryHistory();
    history.push("/1");

    const apiReturnValue = Promise.resolve(mockData);
    getConferenceById.mockImplementation(() => apiReturnValue);

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <Router history={history}>
          <ConferenceDetails />
        </Router>
      );

      await apiReturnValue;
    });
    wrapper.update();
    expect(wrapper.containsMatchingElement(<h3>{mockData.name}</h3>)).toEqual(
      true
    );
    expect(wrapper.containsMatchingElement(<p>{mockData.topic}</p>)).toEqual(
      true
    );
    expect(wrapper.containsMatchingElement(<p>12/11/2019</p>)).toEqual(true);
    expect(wrapper.containsMatchingElement(<p>12:34</p>)).toEqual(true);
    expect(wrapper.containsMatchingElement(<p>{mockData.city}</p>)).toEqual(
      true
    );
    expect(
      wrapper.containsMatchingElement(<p>{mockData.description}</p>)
    ).toEqual(true);
  });
});
