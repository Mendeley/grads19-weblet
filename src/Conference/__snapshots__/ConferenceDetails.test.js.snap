// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConferenceDetails renders conference details 1`] = `
<ConferenceDetails>
  <styled.div>
    <div
      className="sc-AykKC jLKTUb"
    >
      <styled.div>
        <div
          className="sc-AykKE fLIiKq"
        >
          <styled.h3>
            <h3
              className="sc-AykKD eEIGaX"
            >
              Festival of Marketing
            </h3>
          </styled.h3>
          <p>
            Marketing
          </p>
          <p>
            12/11/2019
          </p>
          <p>
            12:34
          </p>
          <p>
            London
          </p>
          <p>
            From Festivalofmarketing.com: The Festival of Marketing is a unique experience where ambitious marketers can discover, learn, celebrate and shape the future together. As the largest global event dedicated to brand marketers, the Festival reflects the very nature of ...
          </p>
          <Styled(Link)
            to="/"
          >
            <Link
              className="sc-AykKF lnrRjo"
              to="/"
            >
              hey I'm a link
            </Link>
          </Styled(Link)>
        </div>
      </styled.div>
    </div>
  </styled.div>
</ConferenceDetails>
`;
